### ISSUE TITLE
Automated test framework

### ISSUE THREAD

**Issue by szanni** [libui-ng/libui-ng#74]
I would love to have some way of running automated tests, preferably some unit testing framework.
I am aware that a lot of things will still have to be tested manually, but at least covering some of the code base in an automated fashion would be an improvement.

I would suggest some type of unit testing framework over `assert()`. I know this most likely would mean introducing an external dependency. We could restrict usage to solely tests though.

Here a [list of unit testing framework on Wikipedia](https://en.wikipedia.org/wiki/List_of_unit_testing_frameworks#C). I would highly recommend a framework with mock support. From personal experience it becomes a pain really fast without.

I personally use cmocka, so my vote goes for that. In the end it doesn't really matter as long as the framework is mature.

**Comment by rubyFeedback** [comment]
Guess you have to ask cody how he feels about unit testing (or other frameworks); and then other contributors. I don't know C well enough to be able to judge myself either way, but I think for any adoption there is some barrier or hurdle to overcome potentially. How may this impact new contributors who just want to quickly add some code/fix? Would they have to write new testing code first before able to send/submit patches or pull requests? Are these testing frameworks exclusive over one another? I guess the ideal would be to be able to retain the simplistic asserts() (if these are used; I did not check but I assume you did since you mentioned it in the report), and then "plug in" more test frameworks if need be or useful. Sort of like a plugin-like structure rather than a "one-size-fits-all" approach. These are just some semi-generic questions how it may impact or change a project. For instance, while testing is important, in ruby I often feel as if I lose too much time adding tests, in particular for simple code. For libraries used by many other people this may be different of course.

**Comment by cody271** [comment]
> I would love to have some way of running automated tests, preferably some unit testing framework.

I'm very supportive of this idea. Testing is always good! The only reason it wasn't on my immediate roadmap is because of how to difficult it actually is in practice. Main issues are:

Even just adding basic unit testing with cmocka requires a lot of work, as libui-ng was written with the assumption of never running headless. Creating a virtual offscreen display within CI is absolutely doable, just very complex. Especially on all three platforms!

Unit testing only gets us so far. The real answer is to use a [graphical testing automation framework](https://en.wikipedia.org/wiki/Comparison_of_GUI_testing_tools). Again, testing on all three platforms with these tools is really quite challenging.

**Comment by szanni** [comment]
I was not even thinking about GUI testing, to my knowledge even projects like _wine_ do not have such things. At least I didn’t see anything like that when submitting patches just recently for a bug that came up when I was testing the table selection PR.

I was thinking more about get/set, new/free and similar things. That would help with memory corruption and leakage, missing implementation, regressions and could probably cover about ~50% of the code.

For the other 50% I was actually thinking about having a side pane in the tester for each component with manual instructions:

1. Click this. Expected result: X
2. Move slider. Expected result: value changes in label
3. …

Even that is a fair amount of work but would at least give reproducible results.
Anyone editing component X must go through these tests on all (available) platforms and things could be tested and easily be confirmed by others (tested and signed off on win/unix etc.)

I'm happy to submit such a proposal when I get back (will not have time for the next 2-3 months) and was more interested in a third party lib even being accepted for testing purposes before I go through all the trouble.

**Comment by parro-it** [comment]
Hi all. @mischnic build [screenshot-tester](https://github.com/mischnic/screenshot-tester) years ago, specifically to test libui-node by comparing windows screenshot against stored version. It proved to be useful, I guess it can be adapted to test libui directly.

**Comment by szanni** [comment]
Closing as resolved: #116

> Hi all. @mischnic build [screenshot-tester](https://github.com/mischnic/screenshot-tester) years ago, specifically to test libui-node by comparing windows screenshot against stored version. It proved to be useful, I guess it can be adapted to test libui directly.

Definitely looks very useful, I believe this should be a different issue though?

**Comment by kojix2** [comment]
Can we automate the taking of screenshots?
I asked AIs.
 
3. ask ChatGPT if there is a way to take screenshots in GitHub Actions with libui-ng.
4. ask ChatGPT to write a prompt to ask DeepWiki.com
5. ask DeepWiki.com to put together a suggestion and take a screenshot in Firefox
6. use Google's OCR to convert it back to a string and have ChatGPT write the plan again.

And got the following results
With vibe coding, we might be able to implement this.

**Comment by kojix2** [comment]
# Proposal: Cross-Platform Screenshot Support for **libui-ng**

Author: ChatGPT

### TL;DR

Add a lightweight, opt-in API that lets CI pipelines (and developers) capture PNG screenshots of `uiWindow` / `uiArea` contents on Windows, macOS, and Linux.
This enables visual regression tests, automated UI documentation, and faster debugging—without pulling in heavyweight graphics stacks like Cairo.

---

## 1 — Why?

| Current Pain-Point                       | Benefit of Screenshots                                 |
| ---------------------------------------- | ------------------------------------------------------ |
| GUI regressions are only caught manually | CI fails instantly when the UI layout/coloring changes |
| Hard to document UI states               | Docs & READMEs can embed freshly-generated images      |
| Reviewing PRs with UI tweaks is slow     | Maintainers see side-by-side diffs in the CI log       |

---

## 2 — Public API (header-only)

```c
/* ui_screenshot.h  (included by ui.h) */
#ifdef LIBUI_CI_SCREENSHOT

_UI_EXTERN bool uiWindowSaveScreenshot(uiWindow *w,
                                       const char *pngFilename);

_UI_EXTERN bool uiAreaSaveScreenshot(uiArea   *a,
                                     const char *pngFilename);
#endif
```

* **Opt-in**: Compiled only when `-DLIBUI_CI_SCREENSHOT` is set.
* **Return value**: `true` on success, `false` on failure (file not written, etc.).

---

## 3 — OS Back-Ends

| Platform        | File                     | Technique                                                                         | External Deps                   |
| --------------- | ------------------------ | --------------------------------------------------------------------------------- | ------------------------------- |
| **Windows**     | `windows/screenshot.cpp` | `WM_PRINTCLIENT` → `HDC` → `GDI+` `Bitmap::Save()`                                | *None* (GDI+ is in OS)          |
| **macOS**       | `darwin/screenshot.m`    | `NSView` → `cacheDisplayInRect:` → `NSBitmapImageRep` → `PNG`                     | *None* (AppKit)                 |
| **Linux / GTK** | `unix/screenshot.c`      | `gtk_widget_draw()` into `cairo_image_surface_t` → `cairo_surface_write_to_png()` | Cairo (already required by GTK) |

> **Why WM\_PRINTCLIENT?**
> libui-ng already implements this message for off-screen rendering of `uiArea`; it works even on headless runners because it doesn’t require the control to be shown on a physical monitor.

---

## 4 — Typical Usage

```c
#include "ui.h"
#include "ui_screenshot.h"

int main(void) {
    uiInitOptions o = {0};
    uiInit(&o);

    uiWindow *w = uiNewWindow("Visual test", 400, 300, 0);
    uiArea   *a = uiNewArea(&someHandler);
    uiWindowSetChild(w, uiControl(a));
    uiWindowShow(w);

    /* --- Take PNGs (only if compiled with -DLIBUI_CI_SCREENSHOT) --- */
#ifdef LIBUI_CI_SCREENSHOT
    uiWindowSaveScreenshot(w, "window.png");
    uiAreaSaveScreenshot(a, "area.png");
#endif

    uiMain();
    uiUninit();
}
```

---

## 5 — Meson Integration

```meson
# meson_options.txt
option('screenshot_tests', type: 'boolean', value: false,
       description: 'Build screenshot back-ends for CI')

# meson.build (snippet)
if get_option('screenshot_tests')
  add_project_arguments('-DLIBUI_CI_SCREENSHOT', language: ['c','cpp','objc'])

  screenshot_sources = []
  os = host_machine.system()
  if os == 'windows'
    screenshot_sources += 'windows/screenshot.cpp'
  elif os == 'darwin'
    screenshot_sources += 'darwin/screenshot.m'
  else
    screenshot_sources += 'unix/screenshot.c'
  endif
  libui_src += screenshot_sources
endif
```

---

## 6 — GitHub Actions Example

```yaml
name: CI + Screenshots

on:
  push:
    branches: [main]
  pull_request:

jobs:
  build-test:
    strategy:
      matrix:
        os: [windows-latest, macos-latest, ubuntu-latest]
    runs-on: ${{ matrix.os }}

    steps:
      - uses: actions/checkout@v4

      - name: Install deps
        if: matrix.os == 'ubuntu-latest'
        run: sudo apt-get update && sudo apt-get install -y ninja-build meson libgtk-3-dev

      - name: Configure
        run: meson setup builddir -Dscreenshot_tests=true

      - name: Build & Tests
        run: |
          meson compile -C builddir
          builddir/test_screenshot      # produces PNGs
          ls builddir/*.png
```

*Artifacts* (PNG files) can be uploaded with `actions/upload-artifact` for manual inspection or diffed automatically via tools like [`pixelmatch`](https://github.com/mapbox/pixelmatch).

---

## 7 — Road-Map

| Phase | Goal                                   | PR contents                         |
| ----- | -------------------------------------- | ----------------------------------- |
| 1     | Windows backend + manual sample        | `windows/screenshot.cpp`            |
| 2     | Add Meson toggle & CI workflow         | Meson changes + GitHub Actions file |
| 3     | macOS backend                          | `darwin/screenshot.m`               |
| 4     | Linux/GTK backend                      | `unix/screenshot.c`                 |
| 5     | Optional pixel-diff regression harness | Separate module / follow-up PR      |

---

## 8 — Maintenance & Safety

* **Zero impact** on default builds (flag-gated).
* Uses *only* OS-provided APIs, so no new runtime dependencies.
* Fails gracefully on headless runners—returns `false` if a surface can’t be created.
